Test installer's raid1 feature

ydirson opened the pull request:
This builds on top of:
- #226 
- #319

ydirson commented:
Last push rebases, and fixes the issues in AnswerFile dict typechecking

glehmann commented on tests/install/conftest.py:
```diff
@@ -98,7 +98,15 @@ def installer_iso(request):
 def system_disks_names(request):
     firmware = request.getfixturevalue("firmware")
-    yield {"uefi": "nvme0n1", "bios": "sda"}[firmware]
+    system_disk_config = request.getfixturevalue("system_disk_config")
```
`system_disk_config` could be validated here.

Maybe use an enum to represent the possible values?

Also `disk` could have a more explicit name like `single_disk`.

ydirson replied:
> `system_disk_config` could be validated here

do you mean type-hinted?

> Also `disk` could have a more explicit name like `single_disk`.

I am not very happy with `disk` but those land in the test name, so I'd prefer a short alternative.  For context, after piling IPv6 parameter on top of this, the test are like:
```
tests/install/test.py::TestNested::test_boot_inst[uefi-83nightly-host1-disk-iso-nosr-ipv6static]
```

And we are not anywhere near what they ought to be :frowning_face:

glehmann replied:
> > `system_disk_config` could be validated here
> 
> do you mean type-hinted?

I mean validated at run time to make sure that the value is a valid one.

An enum could help also with the static analysis. I think that mypy/pyright are able to check when we are not checking all the possible variants

glehmann commented on tests/install/test.py:
```diff
@@ -74,31 +75,47 @@ class TestNested:
             {"TAG": "primary-disk",
              "guest-storage": "no" if local_sr == "nosr" else "yes",
-             "CONTENTS": system_disks_names[0]},
+             "CONTENTS": ("md127" if system_disk_config == "raid1"
```
I'm not a fan of nesting conditional expressions. I think I would prefer something like:

~~~python
dict(raid1='md127', disk=system_disks_names[0])[system_disk_config]
~~~

ydirson replied:
I used those in a few places (I do like getting in a single short line what would otherwise take 4), but in this particular place we'll want to be consistent and not mix the 2 idioms, and I'm not sure how readable that would get with the larger expressions.
OTOH, getting real `ValueError` raised instead of having to add them in the type hints has a lot of appeal... this needs to be looked at.

glehmann replied:
why didn't you raise the `ValueError` in the first place if that's an acceptable option?

I think that

```python
def raise_value_error(v) -> SimpleAnswerfileDict:
    raise ValueError(v)

{"TAG": "raid", "device": "md127",
  "CONTENTS": [
     {"TAG": "disk", "CONTENTS": diskname} for diskname in system_disks_names
 ],
} if system_disk_config == "raid1"
else None if system_disk_config == "disk"
else raise_value_error(f"system_disk_config {system_disk_config!r}")
```

should have worked (not tasted). Am I missing something?

ydirson commented on lib/installer.py:
```diff
@@ -1,42 +1,47 @@
 
-    def top_append(self, *defs):
+    def top_append(self, *defs: Union[SimpleAnswerfileDict, None, ValueError]) -> Self:
+        assert not isinstance(self.defn['CONTENTS'], str), "a toplevel CONTENTS must be a list"
```
The goal of this is to be as close as possible for checking that `self.defn` is indeed of a subclass of its formal type hint. I had a quick try at actually writing a type hint for this but got into problems (which does not mean it is not feasible, I chose concentrated on other things at that time).

`self.defn` is typed `SimpleAnswerfileDict`, so `self.defn['CONTENTS']` is `NotRequired[Union[str, "SimpleAnswerfileDict", Sequence["SimpleAnswerfileDict"]]]`.

The code assumes a "toplevel" is rather typed `Required[Union["SimpleAnswerfileDict", Sequence["SimpleAnswerfileDict"]]]`, so:
- if we stick we `assert` I missed an `assert "CONTENTS" in self.defn` first
- the problem with asserting the actual types is that `isinstance` cannot check using those type hints, which is why I chose to check `self.defn` is just not in the complementary set

But yes, this is a bit hackish and should be better explained in a comment, or simply typed better.

ydirson replied:
It would be more pythonic to keep as much as possible at the type-hint level :)

glehmann commented on lib/typing.py:
```diff
@@ -1,14 +1,39 @@
+    'TAG': str,
+    'CONTENTS': NotRequired[Union[str, "SimpleAnswerfileDict", Sequence["SimpleAnswerfileDict"]]],
+
+    # No way to allow arbitrary fields in addition?  This conveys the
```
yeah, I suppose there's no good solution here

ydirson commented on lib/installer.py:
```diff
@@ -45,29 +50,37 @@ def _normalize_structure(defn):
                 pass            # already copied
             else:
-                new_defn[key] = value
+                new_defn[key] = value # type: ignore[literal-required]
```
> IMO it would have been simpler to keep the old implementation (before the previous commit that prepared this one)

Not sure which commit you mean, its name would help :slightly_smiling_face:

ydirson replied:
> Do we have some benefit to do it that way, with a cast and a type: ignore?

The problem is, it cannot be just a cast.  The original implementation gradually mutates a `SimpleAnswerfileDict` into a `AnswerfileDict`, and the typechecker is not smart enough.

on:
```python
            defn['CONTENTS'] = [AnswerFile._normalize_structure(item)
                                for item in defn['CONTENTS']]
```

pyright:
```
  /home/user/src/xcpng/tests/lib/installer.py:47:13 - error: Could not assign item in TypedDict
    Type "list[AnswerfileDict]" is not assignable to type "str | SimpleAnswerfileDict | Sequence[SimpleAnswerfileDict]"
...
      "list[AnswerfileDict]" is not assignable to "Sequence[SimpleAnswerfileDict]"
        Type parameter "_T_co@Sequence" is covariant, but "AnswerfileDict" is not a subtype of "SimpleAnswerfileDict"
          "CONTENTS" is not required in "SimpleAnswerfileDict"
          "CONTENTS" is an incompatible type
            Type "str | list[AnswerfileDict]" is not assignable to type "str | SimpleAnswerfileDict | Sequence[SimpleAnswerfileDict]"
```

I tried quite a few things and that quickly becomes a nightmare of casts which defeats the type-checking.  Right now I don't see a better way to write `_normalize_structure`, and that `type: ignore` is only for `mypy`, `pyright` does not seem to have any issue there (which might be another issue, I would expect it to require us to duplicate the known-attributes for `AnswerfileDict`).

Anyway, if we move those attributes into a separate dict as discussed in a thread below, the need for that override will vanish.

ydirson commented on lib/installer.py:
```diff
@@ -1,42 +1,47 @@
     @staticmethod
-    def _normalize_structure(defn):
+    def _normalize_structure(defn: Union[SimpleAnswerfileDict, ValueError]) -> AnswerfileDict:
         assert isinstance(defn, dict), f"{defn!r} is not a dict"
```
Well, the whole idea of `ValueError` here is that we'd like to get it thrown in the first place, so we might not need it in the type any more if we switch from ternary operator to dict lookups

ydirson commented:
Separated those preliminar commits not directly tied to the topic, and that did not receive any remark, into #319

rzr commented:
Please rebase

ydirson committed:
installer: typecheck of AnswerFile dict

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
install: add raid1/disk variants

Adds a new system_disk_config test parameter to distinguish raid1
setup from single-disk one.  Valid values for this parameter are
coupled to the product+version, and we also want to limit the test
combinations - from which we explicitly exclude 830 initial iso, in
which answerfile support for installing on raid had been broken.

Adds Alpine-specific setup for tune_firstboot to manually assemble the
RAID.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
install: add support for xsraid1

This allows to check that 8.3 properly works when we update a XS 8.4
installation over upstream swraid.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
install: add a from-scratch reinstall test

Brand new install over a previous installation, to allow checking the
behaviour of pre-existing RAID setups.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

vxgmichel commented on lib/installer.py:
```diff
@@ -6,41 +6,43 @@
     # makes a mutable deep copy of all `contents`
     @staticmethod
-    def _normalize_structure(defn: dict[str, Any]) -> dict[str, Any]:
+    def _normalize_structure(defn: SimpleAnswerfileDict | ValueError) -> AnswerfileDict:
```
I didn't see any use of this `ValueError`, did I miss something?
```suggestion
    def _normalize_structure(defn: SimpleAnswerfileDict) -> AnswerfileDict:
```

ydirson replied:
Right, this could be better documented.  `ValueError` will be generated when expanding an `Answerfile` specification, e.g. when a new `system_disk_config` value is added, and an existing specification does not get updated to handle it, such as [here](https://github.com/xcp-ng/xcp-ng-tests/blob/3217ed6146ed9fe68a3e03bff3f8ab93d09cceaf/tests/install/test.py#L108-L112) (added in this commit, but I fail to see how to link to the change)

vxgmichel commented on lib/installer.py:
```diff
@@ -49,29 +51,39 @@ def _normalize_structure(defn: dict[str, Any]) -> dict[str, Any]:
                 if isinstance(value, str):
                     new_defn['CONTENTS'] = value
                 else:
+                    # TypedDict is not typed enough?
+                    value_as_sequence: Sequence[SimpleAnswerfileDict]
+                    if isinstance(value, Sequence):
+                        value_as_sequence = value
+                    else:
+                        # FIXME what is that?
+                        value_as_sequence = (
+                            cast(SimpleAnswerfileDict, value),
+                        )
```
I think most of the struggle with type inference here is due to the fact that `if key == 'CONTENTS':` is not supported by type checkers as a way to infer the type of `value`.

This could be fixed by using `defn.get('CONTENTS')` instead:
```python
    @staticmethod
    def _normalize_structure(defn: SimpleAnswerfileDict) -> AnswerfileDict:

        # type mutation through nearly-shallow copy
        new_defn: AnswerfileDict = {
            'TAG': defn['TAG'],
            'CONTENTS': [],
        }

        contents = defn.get('CONTENTS')

        if contents is None:
            pass
        elif isinstance(contents, str):
            new_defn['CONTENTS'] = contents
        else:
            value_as_sequence = contents if isinstance(contents, Sequence) else (contents,)
            new_defn['CONTENTS'] = [
                AnswerFile._normalize_structure(item)
                for item in value_as_sequence
                if item is not None
            ]

        for key, value in defn.items():
            if key in ('TAG', 'CONTENTS'):
                continue
            new_defn[key] = value # type: ignore[literal-required]

        return new_defn
```

Side note: this would also be a good use case for `match/case`:
```python3
        match defn.get('CONTENTS'):
            case None:
                pass
            case str() as contents:
                new_defn['CONTENTS'] = contents
            case Sequence() as contents:
                new_defn['CONTENTS'] = [
                    AnswerFile._normalize_structure(item)
                    for item in contents
                    if item is not None
                ]
            case Mapping() as contents:
                new_defn['CONTENTS'] = [AnswerFile._normalize_structure(contents)]
```

The problem is, `mypy` is struggling with inferring that the last `contents` is a `SimpleAnswerfileDict`, something that `pyright` has no problem with. That would be another another reason to stick to `pyright` as discussed [here](https://github.com/xcp-ng/xcp-ng-tests/pull/460#discussion_r3318229852)

vxgmichel commented on lib/installer.py:
```diff
@@ -49,29 +51,39 @@ def _normalize_structure(defn: dict[str, Any]) -> dict[str, Any]:
+        name = defn_copy.pop('TAG')
         assert isinstance(name, str)
-        contents = defn.pop('CONTENTS', ())
+        contents = cast(str | list[AnswerfileDict], defn_copy.pop('CONTENTS', []))
```
The cast here can be removed with a bit of refactoring:
```python
    @staticmethod
    def _defn_to_xml_et(defn: AnswerfileDict) -> ET.Element:
        name = defn['TAG']
        contents = defn['CONTENTS']
        element = ET.Element(name)
        for key, value in defn.items():
            if key in ('CONTENTS', 'TAG'):
                continue
            assert isinstance(value, str)
            element.set(key, value)

        if isinstance(contents, str):
            element.text = contents
        else:
            for content in contents:
                element.append(AnswerFile._defn_to_xml_et(content))
        return element
```

Note that this refactoring also gets rid of the unnecessary `parent` argument.

vxgmichel commented on lib/typing.py:
```diff
@@ -15,3 +17,29 @@
+# - does not require to write 0 or 1 subelement as a list
+SimpleAnswerfileDict = TypedDict('SimpleAnswerfileDict', {
+    'TAG': str,
+    'CONTENTS': NotRequired[Union[str, "SimpleAnswerfileDict", Sequence["SimpleAnswerfileDict"]]],
```
I'd suggest to consistently use `|` instead of `Union` in the code base, as `Union` is the pre-[PEP 604](https://peps.python.org/pep-0604/) version:
```suggestion
    'CONTENTS': NotRequired[str | "SimpleAnswerfileDict" | Sequence["SimpleAnswerfileDict"]],
```
See the [python docs](https://docs.python.org/3/library/typing.html#typing.Union):
> To define a union, use e.g. Union[int, str] or the shorthand int | str. Using that shorthand is recommended.

Note that it's also how type checkers choose to display their types:
```python3
def f(x: Union[int, str]) -> None: reveal_type(x)  # information: Type of "x" is "int | str"
```

ydirson replied:
Yes, that code predates the bump of required python version to support this syntax

vxgmichel commented on tests/install/test-sequences/inst+upg+rst.lst:
```diff
@@ -1,2 +1,2 @@
-tests/install/test.py::TestNested::test_restore[uefi-83nightly-83nightly-83nightly-iso-ext]
-tests/install/test.py::TestNested::test_boot_rst[uefi-83nightly-83nightly-83nightly-iso-ext]
+tests/install/test.py::TestNested::test_restore[uefi-83nightly-disk-83nightly-83nightly-iso-ext]
+tests/install/test.py::TestNested::test_boot_rst[uefi-83nightly-disk-83nightly-83nightly-iso-ext]
```
Like @glehmann, I find `disk` confusing in this context. Since you mentioned you don't want to it to be too long (as in `single-disk` for instance), I'd suggest `nonraid`  which is the same width as `xsraid1`.

While looking for other alternatives, I found that [JBOD](https://en.wikipedia.org/wiki/Non-RAID_drive_architectures#JBOD) ("Just a Bunch Of Disks") is apparently a thing. I don't know how well this acronym is known though.

ydirson replied:
I agree it's really not a good naming.

JBOD is actually implying several disks (which can even be combined in RAID) so I don't think it would be a good fit.

One idea I had was "plain disk" but that's long and "plain" alone would not be self-explanatory.

Another aspect I thought could be factored in is that we're actually testing on different disk types (scsi and nvme, though they are actually artificially tied to bios/uefi choice by XAPI), and we will want at some point to include installation over networked targets (iSCSI, HBA/FC as described in [XS doc](https://docs.xenserver.com/en-us/xenserver/8/install/advanced-install#boot-from-san)).
As such I thought we can have in the future choices like scsi/nvme/scsi+raid1/nvme+xsraid1/iscsi/hba, as anyway raid setups don't usually make sense on boot-from-san setups where data redundancy would be the SAN's job.

So what we need would be one of:
- find a wording for "local disk" that doesn't suck, but then if/when scsi/nvme ever gets decoupled from bios/uefi we'd have to make another naming change
- use scsi/nvme/scsi+raid1/nvme+xsraid1 but couple them with uefi/bios; impact would be swapping this parameter with the "iso version" one, slightly breaking the "order of params is order of choices done" I've tried to keep in the parameter ordering, but arguably you'll make that choice before booting anyway

I guess the latter actually makes more sense

vxgmichel commented on tests/install/test.py:
```diff
@@ -78,25 +81,45 @@ class TestNested:
+                  + ([dict(name="vm1 system disk mirror", size="100GiB", device="xvdb", userdevice="1")]
+                     if system_disk_config in ("raid1", "xsraid1") else [])
+                  ),
             cd_vbd=dict(device="xvdd", userdevice="3"),
             vifs=[dict(index=0, network_name=NETWORKS["MGMT"])],
         ))
     @pytest.mark.answerfile.with_args(
-        lambda system_disks_names, local_sr, package_source, iso_version: AnswerFile("INSTALL")
+        lambda system_disks_names, local_sr, package_source, system_disk_config, iso_version: AnswerFile("INSTALL")
         .top_setattr({} if local_sr == "nosr" else {"sr-type": local_sr})
         .top_append(
             {"iso": {"TAG": "source", "type": "local"},
              "net": {"TAG": "source", "type": "url",
                      "CONTENTS": ISO_IMAGES[iso_version]['net-url']},  # type: ignore
              }[package_source],
+
+            {"raid1": SimpleAnswerfileDict(
+                {"TAG": "raid", "device": "md127",
+                 "CONTENTS": [
+                     {"TAG": "disk", "CONTENTS": diskname} for diskname in system_disks_names
+                 ]}),
+             "disk": None,
+             "xsraid1": None,
+             }[system_disk_config],
+
             {"TAG": "admin-interface", "name": "eth0", "proto": "dhcp"},
             {"TAG": "primary-disk",
              "guest-storage": "no" if local_sr == "nosr" else "yes",
-             "CONTENTS": system_disks_names[0]},
+             "swraid": "yes" if system_disk_config == "xsraid1" else "no",
+             "CONTENTS": {"disk": system_disks_names[0],
+                          "raid1": "md127",
+                          "xsraid1": ','.join(system_disks_names),
+                          }[system_disk_config],
+             },
         ))
```
I think this large lambda would be better as a `make_answer_file(system_disks_names, local_sr, package_source, system_disk_config, iso_version)` function close to the `AnswerFile` definition (or a class method).

This way, the answer file logic is not spread across several modules. The function could also be refactored into several methods such as `AnswerFile.append_package_source(...)`, `AnswerFile.append_system_disk_config(...)`, etc. 

I think it could help future readers to better understand how answerfile are structured.

It would also avoid code duplication with the similar lambda above `test_reinstall` and `test_upgrade`

vxgmichel commented on tests/install/test.py:
```diff
@@ -148,12 +197,29 @@ def _test_firstboot(self, create_vms: list[VM], mode: str, *,
-            expected_rel_id = split_mode[-3]
+            assert len(split_installed_config) >= 4, ("unexpectedly-short installed_config"
+                                                      f" {installed_config!r} for restore")
+            # FIXME: this is an awful heuristic
+            if len(split_installed_config) == 4:
+                # reverting the first upgrade, must skip system_disk_config
+                expected_rel_id = split_installed_config[-4]
+            else:
+                expected_rel_id = split_installed_config[-3]
```
Shouldn't this `FIXME` be addressed before merging?

vxgmichel commented on tests/install/test.py:
```diff
@@ -105,27 +128,53 @@ def test_install(self, vm_booted_with_installer: VM, system_disks_names: list[st
+            case "disk":
+                helper_vm.ssh("mount /dev/xvdb1 /mnt")
+            case "raid1" | "xsraid1":
+                # FIXME helper VM has to be an Alpine, that should not be a random vm_ref
+                logging.info("Assembling RAID (%s)", system_disk_config)
+                helper_vm.ssh("apk add mdadm")
+                helper_vm.ssh("mdadm -A --scan")
+                helper_vm.ssh("mount /dev/md127p1 /mnt")
```
Shouldn't this `FIXME` be addressed before merging?

ydirson replied:
That's linked to how the XVA for VM import is selected (using the existing flexible mechanism), so the way to go today would be I guess to add yet another marker like we did for `debian_uefi_vm`, though that mostly highlights the limitations of the current approach (which also bleeds on us with the `None` parameter in `tune_first_boot`)

vxgmichel commented on data.py-dist:
```diff
@@ -240,26 +240,26 @@ LVMOHBA_DEVICE_CONFIG: dict[str, str] = {
-            {"TAG": "keymap",
-             "CONTENTS": "us"},
+            SimpleAnswerfileDict({"TAG": "root-password",
+                                  "type": "hash",
+                                  "CONTENTS": HOST_DEFAULT_PASSWORD_HASH}),
+            SimpleAnswerfileDict({"TAG": "timezone",
+                                  "CONTENTS": "Europe/Paris"}),
+            SimpleAnswerfileDict({"TAG": "keymap",
+                                  "CONTENTS": "us"}),
         ),
-    },
-    UPGRADE={
+    }),
+    UPGRADE=SimpleAnswerfileDict({
         "TAG": "installation",
         "mode": "upgrade",
-    },
-    RESTORE={
+    }),
+    RESTORE=SimpleAnswerfileDict({
         "TAG": "restore",
-    },
+    }),
 )
```
Maybe that's not the right place to discuss this, but most of this structure is actually not configuration (for instance configuring different tags would not make sense).

Wouldn't be better to have the following:
```python
ANSWER_FILE_INSTALL_ROOT_PASSWORD_HASH = HOST_DEFAULT_PASSWORD_HASH
ANSWER_FILE_INSTALL_TIMEZONE = "Europe/Paris"
ANSWER_FILE_INSTALL_KEYMAP = "us"
# Add more when needed
```

ydirson replied:
Yes that part has been in need to be reworked for ages. My initial idea was still to keep the addition of the "installation" contents as a whole, but actually selecting the individual items like that looks like a good idea.  Maybe we should do that as a preliminary PR.

vxgmichel reviewed (changes requested):
Most of my comments are suggestions, but I think a couple of them still require changes (such as the seemingly unused `ValueError` argument in the `_normalize_structure` methods. 

LGTM otherwise 👍
